ufl/algorithms/apply_derivatives.py

```diff
@@ -562,6 +563,14 @@ def __init__(self, geometric_dimension):
         """Initialise."""
         GenericDerivativeRuleset.__init__(self, var_shape=(geometric_dimension,))
         self._Id = Identity(geometric_dimension)
+        self.degree_estimator = SumDegreeEstimator(1, {})
+
+    def is_cellwise_constant(self, o):
+        """More precise checks for cellwise constants."""
+        if is_cellwise_constant(o):
+            return True
+        degree = map_expr_dag(self.degree_estimator, o)
```

Comment:
You've introduced here that `is_cellwise_constant` logic is based on degree estimation, which is the only place in the code. How does it work, since the degree estimation for geometric quantities calls `is_cellwise_constant` itself?   https://github.com/FEniCS/ufl/blob/1ab30c2de2a975b9fc957c6b356541bd15b35e15/ufl/algorithms/estimate_degrees.py#L46-L56

If there is no infinite loop, could this be used in the generic cell-wise constant check? Or is it that only `GradRuleset` does not cause infinite loop?

Reply:
I first tried to add this logic to the generic `is_cellwise_constant` in https://github.com/FEniCS/ufl/pull/334, but simplifying `grad(cell_wise_constant)` on instantiation turns out to be problematic for nested `grad` expression simply because `grad(f)` expects f to have a domain and `Zero` does not have a domain, so  `grad(grad(grad(linear)))` would simplify to `grad(Zero)`.